File: exercism.io/ruby/word-count/word_count.rb
class Phrase
  def initialize(phrase)
    @phrase = phrase
  end

  def word_count
    phrase
      .downcase
      .gsub(/[^\w']/, ' ')
      .split(' ')
      .group_by { |word| word.gsub(/'([^']+)'/, '\1') }
      .reduce({}) { |acc, (word, list)| acc.merge!(word => list.size) }
  end

  private

  attr_reader :phrase
end

File: exercism.io/elixir/word-count/word_count.exs
defmodule Words do
  @doc """
  Count the number of words in the sentence.

  Words are compared case-insensitively.
  """
  @spec count(String.t()) :: map
  def count(sentence) do
    sentence
    |> sanitize()
    |> String.split(" ", trim: true)
    |> Enum.reduce(%{}, &count/2)
  end

  defp sanitize(sentence) do
    sentence
    |> String.downcase()
    |> String.replace(~r/[^\d\p{L}-]/u, " ")
  end

  defp count(word, counter) do
    Map.merge(counter, %{word => Map.get(counter, word, 0) + 1})
  end
end
